File: scripts/data_processing/01_data_cleaning.py
"""
============================================
NYC Taxi Fairness Audit - Data Cleaning
============================================
Stage 3A: Clean raw taxi data using PySpark

This script:
1. Loads raw taxi data from HDFS/local
2. Removes missing values
3. Removes outliers
4. Calculates derived fields
5. Saves cleaned data as Parquet

Author: Big Data Analytics Project
"""

from pyspark.sql import SparkSession
from pyspark.sql.functions import (
    col, when, hour, dayofweek, unix_timestamp,
    lit, monotonically_increasing_id, concat, 
    round as spark_round, abs as spark_abs
)
from pyspark.sql.types import *
import os
import sys

# ============================================
# CONFIGURATION
# ============================================

# For local testing (without HDFS)
LOCAL_MODE = True

if LOCAL_MODE:
    INPUT_PATH = "data/yellow_tripdata_*.csv"
    OUTPUT_PATH = "output/processed/taxi_cleaned"
    CENSUS_PATH = "data/us_income_zipcode.csv"
else:
    INPUT_PATH = "hdfs:///bigdata/taxi/raw/*.csv"
    OUTPUT_PATH = "hdfs:///bigdata/processed/taxi_cleaned"
    CENSUS_PATH = "hdfs:///bigdata/census/us_income_zipcode.csv"

# Data quality thresholds
MAX_FARE = 500  # Maximum reasonable fare
MAX_DISTANCE = 100  # Maximum reasonable distance (miles)
MAX_DURATION = 300  # Maximum trip duration (minutes)
MAX_SPEED = 100  # Maximum reasonable speed (mph)
MIN_FARE = 2.5  # Minimum fare (base fare)

# ============================================
# SPARK SESSION
# ============================================

def create_spark_session():
    """Create and configure Spark session."""
    spark = SparkSession.builder \
        .appName("NYC_Taxi_Fairness_DataCleaning") \
        .config("spark.sql.parquet.compression.codec", "snappy") \
        .config("spark.sql.shuffle.partitions", "200") \
        .getOrCreate()
    
    spark.sparkContext.setLogLevel("WARN")
    return spark


# ============================================
# DATA LOADING
# ============================================

def load_taxi_data(spark, input_path):
    """Load raw taxi data from CSV files."""
    print(f"\n[1/5] Loading taxi data from: {input_path}")
    
    schema = StructType([
        StructField("VendorID", IntegerType(), True),
        StructField("tpep_pickup_datetime", TimestampType(), True),
        StructField("tpep_dropoff_datetime", TimestampType(), True),
        StructField("passenger_count", IntegerType(), True),
        StructField("trip_distance", DoubleType(), True),
        StructField("pickup_longitude", DoubleType(), True),
        StructField("pickup_latitude", DoubleType(), True),
        StructField("RatecodeID", IntegerType(), True),
        StructField("store_and_fwd_flag", StringType(), True),
        StructField("dropoff_longitude", DoubleType(), True),
        StructField("dropoff_latitude", DoubleType(), True),
        StructField("payment_type", IntegerType(), True),
        StructField("fare_amount", DoubleType(), True),
        StructField("extra", DoubleType(), True),
        StructField("mta_tax", DoubleType(), True),
        StructField("tip_amount", DoubleType(), True),
        StructField("tolls_amount", DoubleType(), True),
        StructField("improvement_surcharge", DoubleType(), True),
        StructField("total_amount", DoubleType(), True)
    ])
    
    df = spark.read.csv(
        input_path,
        header=True,
        schema=schema,
        timestampFormat="yyyy-MM-dd HH:mm:ss"
    )
    
    initial_count = df.count()
    print(f"   Loaded {initial_count:,} records")
    
    return df, initial_count


# ============================================
# DATA CLEANING FUNCTIONS
# ============================================

def remove_missing_values(df):
    """Remove rows with critical missing values."""
    print("\n[2/5] Removing missing values...")
    
    # Define critical columns that cannot be null
    critical_columns = [
        "tpep_pickup_datetime",
        "tpep_dropoff_datetime",
        "pickup_longitude",
        "pickup_latitude",
        "dropoff_longitude",
        "dropoff_latitude",
        "trip_distance",
        "fare_amount"
    ]
    
    # Remove nulls
    df_clean = df.dropna(subset=critical_columns)
    
    # Remove zero coordinates (invalid GPS)
    df_clean = df_clean.filter(
        (col("pickup_longitude") != 0) & 
        (col("pickup_latitude") != 0) &
        (col("dropoff_longitude") != 0) & 
        (col("dropoff_latitude") != 0)
    )
    
    cleaned_count = df_clean.count()
    print(f"   Records after removing nulls: {cleaned_count:,}")
    
    return df_clean


def remove_outliers(df):
    """Remove statistical outliers and invalid data."""
    print("\n[3/5] Removing outliers...")
    
    # Calculate trip duration in minutes
    df = df.withColumn(
        "trip_duration_minutes",
        (unix_timestamp("tpep_dropoff_datetime") - 
         unix_timestamp("tpep_pickup_datetime")) / 60
    )
    
    # Calculate speed (mph)
    df = df.withColumn(
        "speed_mph",
        when(col("trip_duration_minutes") > 0,
             col("trip_distance") / (col("trip_duration_minutes") / 60)
        ).otherwise(0)
    )
    
    # Apply filters for valid trips
    df_clean = df.filter(
        # Valid fare range
        (col("fare_amount") >= MIN_FARE) &
        (col("fare_amount") <= MAX_FARE) &
        # Valid distance
        (col("trip_distance") > 0) &
        (col("trip_distance") <= MAX_DISTANCE) &
        # Valid duration
        (col("trip_duration_minutes") > 0) &
        (col("trip_duration_minutes") <= MAX_DURATION) &
        # Valid speed (not teleporting!)
        (col("speed_mph") <= MAX_SPEED) &
        # Valid passenger count
        (col("passenger_count") > 0) &
        (col("passenger_count") <= 9) &
        # NYC bounding box (roughly)
        (col("pickup_longitude").between(-74.5, -73.5)) &
        (col("pickup_latitude").between(40.4, 41.0)) &
        (col("dropoff_longitude").between(-74.5, -73.5)) &
        (col("dropoff_latitude").between(40.4, 41.0))
    )
    
    # Drop speed column (only needed for filtering)
    df_clean = df_clean.drop("speed_mph")
    
    cleaned_count = df_clean.count()
    print(f"   Records after removing outliers: {cleaned_count:,}")
    
    return df_clean


def add_derived_fields(df):
    """Calculate derived fields for analysis."""
    print("\n[4/5] Adding derived fields...")
    
    # Add unique trip ID
    df = df.withColumn(
        "trip_id",
        concat(
            col("VendorID").cast("string"),
            lit("_"),
            monotonically_increasing_id().cast("string")
        )
    )
    
    # Extract time-based features
    df = df.withColumn("hour_of_day", hour("tpep_pickup_datetime"))
    df = df.withColumn("day_of_week", dayofweek("tpep_pickup_datetime"))
    
    # Rush hour indicator (7-9 AM or 5-7 PM on weekdays)
    df = df.withColumn(
        "is_rush_hour",
        when(
            ((col("hour_of_day").between(7, 9)) | 
             (col("hour_of_day").between(17, 19))) &
            (col("day_of_week").between(2, 6)),  # Monday=2 to Friday=6 in Spark
            True
        ).otherwise(False)
    )
    
    # Weekend indicator
    df = df.withColumn(
        "is_weekend",
        when(
            (col("day_of_week") == 1) | (col("day_of_week") == 7),
            True
        ).otherwise(False)
    )
    
    # Night trip indicator (10 PM - 6 AM)
    df = df.withColumn(
        "is_night",
        when(
            (col("hour_of_day") >= 22) | (col("hour_of_day") <= 6),
            True
        ).otherwise(False)
    )
    
    # Fare per mile (price efficiency)
    df = df.withColumn(
        "fare_per_mile",
        spark_round(col("fare_amount") / col("trip_distance"), 2)
    )
    
    # Round trip duration
    df = df.withColumn(
        "trip_duration_minutes",
        spark_round(col("trip_duration_minutes"), 2)
    )
    
    print("   Added: trip_id, hour_of_day, day_of_week, is_rush_hour, is_weekend, is_night, fare_per_mile")
    
    return df


def assign_zip_codes(df):
    """
    Assign ZIP codes based on pickup coordinates.
    In production, use reverse geocoding or spatial join with NYC ZIP boundaries.
    This is a simplified approximation using coordinate binning.
    """
    print("   Assigning approximate ZIP codes from coordinates...")
    
    # Create approximate ZIP code zones based on latitude/longitude
    # This is a simplified approach - in production, use proper geocoding
    df = df.withColumn(
        "pickup_zone",
        concat(
            spark_round(col("pickup_latitude"), 2).cast("string"),
            lit("_"),
            spark_round(col("pickup_longitude"), 2).cast("string")
        )
    )
    
    return df


# ============================================
# MAIN EXECUTION
# ============================================

def main():
    """Main execution function."""
    print("=" * 60)
    print("NYC TAXI FAIRNESS AUDIT - DATA CLEANING")
    print("=" * 60)
    
    # Create Spark session
    spark = create_spark_session()
    
    try:
        # Load data
        df, initial_count = load_taxi_data(spark, INPUT_PATH)
        
        # Clean data
        df = remove_missing_values(df)
        df = remove_outliers(df)
        df = add_derived_fields(df)
        df = assign_zip_codes(df)
        
        # Select final columns
        final_columns = [
            "trip_id",
            "tpep_pickup_datetime",
            "tpep_dropoff_datetime",
            "passenger_count",
            "trip_distance",
            "pickup_longitude",
            "pickup_latitude",
            "dropoff_longitude",
            "dropoff_latitude",
            "fare_amount",
            "total_amount",
            "trip_duration_minutes",
            "hour_of_day",
            "day_of_week",
            "is_rush_hour",
            "is_weekend",
            "is_night",
            "fare_per_mile",
            "pickup_zone"
        ]
        
        df_final = df.select(final_columns)
        
        # Save cleaned data
        print(f"\n[5/5] Saving cleaned data to: {OUTPUT_PATH}")
        
        df_final.write \
            .mode("overwrite") \
            .parquet(OUTPUT_PATH)
        
        # Print summary
        final_count = df_final.count()
        removed_pct = ((initial_count - final_count) / initial_count) * 100
        
        print("\n" + "=" * 60)
        print("DATA CLEANING COMPLETE")
        print("=" * 60)
        print(f"Initial records:  {initial_count:,}")
        print(f"Final records:    {final_count:,}")
        print(f"Records removed:  {initial_count - final_count:,} ({removed_pct:.1f}%)")
        print(f"Output location:  {OUTPUT_PATH}")
        print("=" * 60)
        
        # Show sample
        print("\nSample of cleaned data:")
        df_final.show(5, truncate=False)
        
        # Print schema
        print("\nFinal schema:")
        df_final.printSchema()
        
    finally:
        spark.stop()


if __name__ == "__main__":
    main()

File: scripts/scala/01_DataCleaning.scala
// ============================================================
// NYC Taxi Fairness Audit  –  Stage 1: Data Cleaning
// ============================================================
// Big Data Frameworks: Apache Spark (Scala), Spark SQL, HDFS
//
// Run:
//   spark-shell --driver-memory 8g -i scripts/scala/01_DataCleaning.scala
//
// Input:  Local CSV  data/*.csv  OR  HDFS  /bigdata/taxi/raw/*.csv
// Output: Parquet    output/processed/taxi_cleaned
//         (also written to HDFS /bigdata/processed/taxi_cleaned when cluster mode)
//
// This script:
//   1. Loads 47M+ raw taxi records from CSV (with explicit schema)
//   2. Drops rows with NULL critical fields
//   3. Filters outliers (fare, distance, duration, speed, GPS bounding box)
//   4. Derives temporal features: hour, day-of-week, rush-hour, weekend, night
//   5. Computes trip_duration_minutes, fare_per_mile, speed_mph
//   6. Assigns a deterministic trip_id (sha2 hash of key fields)
//   7. Writes Snappy-compressed Parquet for downstream Spark & Hive
// ============================================================

import org.apache.spark.sql.{SparkSession, DataFrame}
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._

println("\n" + "=" * 70)
println("  NYC TAXI FAIRNESS AUDIT – DATA CLEANING (Scala / Spark SQL)")
println("  Spark " + spark.version + "  |  Scala " + util.Properties.versionString)
println("=" * 70)

// ──────────────── Configuration ────────────────
val LOCAL_MODE     = true
val INPUT_PATH     = if (LOCAL_MODE) "data/yellow_tripdata_*.csv"
                     else "hdfs:///bigdata/taxi/raw/*.csv"
val OUTPUT_PATH    = if (LOCAL_MODE) "output/processed/taxi_cleaned"
                     else "hdfs:///bigdata/processed/taxi_cleaned"

// Data-quality thresholds
val MAX_FARE       = 500.0    // USD
val MIN_FARE       = 2.50     // NYC base fare
val MAX_DISTANCE   = 100.0    // miles
val MAX_DURATION   = 300.0    // minutes (5 hours)
val MAX_SPEED      = 100.0    // mph
val NYC_LAT_MIN    = 40.49;  val NYC_LAT_MAX = 40.92
val NYC_LON_MIN    = -74.27; val NYC_LON_MAX = -73.68

// ──────────────── Step 1: Load raw CSV with explicit schema ────────────────
println("\n[1/5] Loading raw taxi CSV data...")

val rawSchema = StructType(Array(
  StructField("VendorID",              IntegerType,   true),
  StructField("tpep_pickup_datetime",  TimestampType, true),
  StructField("tpep_dropoff_datetime", TimestampType, true),
  StructField("passenger_count",       IntegerType,   true),
  StructField("trip_distance",         DoubleType,    true),
  StructField("pickup_longitude",      DoubleType,    true),
  StructField("pickup_latitude",       DoubleType,    true),
  StructField("RatecodeID",            IntegerType,   true),
  StructField("store_and_fwd_flag",    StringType,    true),
  StructField("dropoff_longitude",     DoubleType,    true),
  StructField("dropoff_latitude",      DoubleType,    true),
  StructField("payment_type",          IntegerType,   true),
  StructField("fare_amount",           DoubleType,    true),
  StructField("extra",                 DoubleType,    true),
  StructField("mta_tax",               DoubleType,    true),
  StructField("tip_amount",            DoubleType,    true),
  StructField("tolls_amount",          DoubleType,    true),
  StructField("improvement_surcharge", DoubleType,    true),
  StructField("total_amount",          DoubleType,    true)
))

val rawDF = spark.read
  .option("header", "true")
  .option("timestampFormat", "yyyy-MM-dd HH:mm:ss")
  .schema(rawSchema)
  .csv(INPUT_PATH)

val initialCount = rawDF.count()
println(f"   Loaded $initialCount%,d raw records")

// ──────────────── Step 2: Remove NULL critical fields ────────────────
println("\n[2/5] Removing rows with NULL critical fields...")

val criticalCols = Seq("tpep_pickup_datetime", "tpep_dropoff_datetime",
  "pickup_longitude", "pickup_latitude", "dropoff_longitude", "dropoff_latitude",
  "trip_distance", "fare_amount")

var cleanDF = rawDF.na.drop(criticalCols)

// Remove zero GPS (invalid sensor readings)
cleanDF = cleanDF.filter(
  col("pickup_longitude")  =!= 0.0 && col("pickup_latitude")  =!= 0.0 &&
  col("dropoff_longitude") =!= 0.0 && col("dropoff_latitude") =!= 0.0
)

val afterNullCount = cleanDF.count()
println(f"   After NULL removal: $afterNullCount%,d  (dropped ${initialCount - afterNullCount}%,d)")

// ──────────────── Step 3: Derive computed columns ────────────────
println("\n[3/5] Computing derived fields...")

cleanDF = cleanDF
  // Trip duration in minutes
  .withColumn("trip_duration_minutes",
    round((unix_timestamp(col("tpep_dropoff_datetime")) -
           unix_timestamp(col("tpep_pickup_datetime"))) / 60.0, 2))
  // Average speed (mph)
  .withColumn("speed_mph",
    when(col("trip_duration_minutes") > 0,
      round(col("trip_distance") / (col("trip_duration_minutes") / 60.0), 2))
    .otherwise(0.0))
  // Temporal features
  .withColumn("hour_of_day", hour(col("tpep_pickup_datetime")))
  .withColumn("day_of_week", dayofweek(col("tpep_pickup_datetime")))
  // Rush hour: 7-9 AM or 5-7 PM on weekdays (Mon=2..Fri=6 in Spark)
  .withColumn("is_rush_hour",
    ((col("hour_of_day").between(7, 9) || col("hour_of_day").between(17, 19)) &&
      col("day_of_week").between(2, 6)))
  // Weekend: Saturday=7, Sunday=1 in Spark
  .withColumn("is_weekend", col("day_of_week").isin(1, 7))
  // Night: 10 PM - 6 AM
  .withColumn("is_night", col("hour_of_day") >= 22 || col("hour_of_day") <= 5)
  // Fare per mile
  .withColumn("fare_per_mile",
    round(col("fare_amount") / col("trip_distance"), 2))
  // Grid-based pickup zone (0.01° ≈ 1.1 km)
  .withColumn("pickup_zone",
    concat(bround(col("pickup_latitude"), 2).cast("string"), lit("_"),
           bround(col("pickup_longitude"), 2).cast("string")))
  // Deterministic trip_id = SHA-256 of key fields → first 16 hex chars
  .withColumn("trip_id",
    substring(sha2(concat_ws("|",
      col("tpep_pickup_datetime").cast("string"),
      col("pickup_latitude").cast("string"),
      col("pickup_longitude").cast("string"),
      col("fare_amount").cast("string"),
      col("trip_distance").cast("string")), 256), 1, 16))

println("   Added: trip_id, trip_duration_minutes, speed_mph, hour_of_day, " +
        "day_of_week, is_rush_hour, is_weekend, is_night, fare_per_mile, pickup_zone")

// ──────────────── Step 4: Filter outliers ────────────────
println("\n[4/5] Filtering outliers...")

cleanDF = cleanDF.filter(
  // Fare range
  col("fare_amount").between(MIN_FARE, MAX_FARE) &&
  // Distance range
  col("trip_distance") > 0 && col("trip_distance") <= MAX_DISTANCE &&
  // Duration range
  col("trip_duration_minutes") > 0 && col("trip_duration_minutes") <= MAX_DURATION &&
  // Speed sanity
  col("speed_mph") <= MAX_SPEED &&
  // Passengers
  col("passenger_count") > 0 && col("passenger_count") <= 9 &&
  // NYC bounding box
  col("pickup_latitude").between(NYC_LAT_MIN, NYC_LAT_MAX) &&
  col("pickup_longitude").between(NYC_LON_MIN, NYC_LON_MAX) &&
  col("dropoff_latitude").between(NYC_LAT_MIN, NYC_LAT_MAX) &&
  col("dropoff_longitude").between(NYC_LON_MIN, NYC_LON_MAX)
)

val finalCount = cleanDF.count()
val removedPct = ((initialCount - finalCount).toDouble / initialCount * 100)
println(f"   After outlier removal: $finalCount%,d")
println(f"   Total removed: ${initialCount - finalCount}%,d ($removedPct%.1f%%)")

// ──────────────── Step 5: Select final columns & write Parquet ────────────────
println(s"\n[5/5] Writing Parquet to: $OUTPUT_PATH")

val finalColumns = Seq(
  "trip_id",
  "tpep_pickup_datetime", "tpep_dropoff_datetime",
  "passenger_count", "trip_distance",
  "pickup_longitude", "pickup_latitude",
  "dropoff_longitude", "dropoff_latitude",
  "fare_amount", "total_amount",
  "trip_duration_minutes", "speed_mph",
  "hour_of_day", "day_of_week",
  "is_rush_hour", "is_weekend", "is_night",
  "fare_per_mile", "pickup_zone"
)

val outputDF = cleanDF.select(finalColumns.map(col): _*)

outputDF.write
  .mode("overwrite")
  .option("compression", "snappy")
  .parquet(OUTPUT_PATH)

// ──────────────── Summary ────────────────
println("\n" + "=" * 70)
println("  DATA CLEANING COMPLETE")
println("=" * 70)
println(f"  Initial records:  $initialCount%,d")
println(f"  Final records:    $finalCount%,d")
println(f"  Records removed:  ${initialCount - finalCount}%,d ($removedPct%.1f%%)")
println(f"  Output:           $OUTPUT_PATH")
println()

println("  Data Quality Filters Applied:")
println(f"    Fare range:     $$${MIN_FARE}%.2f – $$${MAX_FARE}%.0f")
println(f"    Distance range: 0 – $MAX_DISTANCE%.0f miles")
println(f"    Duration range: 0 – $MAX_DURATION%.0f minutes")
println(f"    Max speed:      $MAX_SPEED%.0f mph")
println(f"    GPS bounding:   NYC metro area")
println()

println("  Schema:")
outputDF.printSchema()

println("  Sample data:")
outputDF.show(5, truncate = false)

println("  Fare distribution:")
outputDF.describe("fare_amount", "trip_distance", "trip_duration_minutes", "speed_mph")
  .show()

println("=" * 70)
println("  Next: Run 02_DataEnrichment.scala")
println("=" * 70)

System.exit(0)
